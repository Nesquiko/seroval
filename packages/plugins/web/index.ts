export { default as BlobPlugin } from './blob';
export { default as FilePlugin } from './file';
export { default as HeadersPlugin } from './headers';
export { default as URLSearchParamsPlugin } from './url-search-params';
export { default as URL } from './url';
