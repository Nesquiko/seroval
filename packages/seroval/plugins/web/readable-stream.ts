import type { SerovalNode, Stream } from 'seroval';
import { createPlugin, createStream } from 'seroval';

function toStream<T>(value: ReadableStream<T>): Stream<T | undefined> {
  const stream = createStream<T | undefined>();

  const reader = value.getReader();

  async function push(): Promise<void> {
    try {
      const result = await reader.read();
      if (result.done) {
        stream.return(result.value);
      } else {
        stream.next(result.value);
        await push();
      }
    } catch (error) {
      stream.throw(error);
    }
  }

  push().catch(() => {
    //
  });

  return stream;
}

interface ReadableStreamNode {
  stream: SerovalNode;
}

export default createPlugin<ReadableStream, ReadableStreamNode>({
  tag: 'seroval/plugins/web/readable-stream',
  test(value) {
    if (typeof ReadableStream === 'undefined') {
      return false;
    }
    return value instanceof ReadableStream;
  },
  parse: {
    sync(value, ctx, data) {
      return {
        stream: ctx.parse(createStream()),
      };
    },
    async async(value, ctx, data) {
      return {
        stream: await ctx.parse(toStream(value)),
      };
    },
    stream(value, ctx, data) {
      return {
        stream: ctx.parse(toStream(value)),
      };
    },
  },
  serialize(node, ctx, data) {

  },
  deserialize(node, ctx, data) {

  },
});
