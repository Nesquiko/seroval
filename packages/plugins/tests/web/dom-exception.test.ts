import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import { DOMExceptionPlugin } from '../../web';

const EXAMPLE_MESSAGE = 'This is an example message.';
const EXAMPLE_NAME = 'Example';
const EXAMPLE = new DOMException(EXAMPLE_MESSAGE, EXAMPLE_NAME);

describe('DOMException', () => {
  describe('serialize', () => {
    it('supports DOMException', () => {
      const result = serialize(EXAMPLE, {
        plugins: [DOMExceptionPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(DOMException);
      expect(back.message).toBe(EXAMPLE.message);
      expect(back.name).toBe(EXAMPLE.name);
    });
  });
  describe('serializeAsync', () => {
    it('supports DOMException', async () => {
      const result = await serializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [DOMExceptionPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof EXAMPLE>>(result);
      expect(back).toBeInstanceOf(DOMException);
      expect(back.message).toBe(EXAMPLE.message);
      expect(back.name).toBe(EXAMPLE.name);
    });
  });
  describe('toJSON', () => {
    it('supports DOMException', () => {
      const result = toJSON(EXAMPLE, {
        plugins: [DOMExceptionPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result, {
        plugins: [DOMExceptionPlugin],
      });
      expect(back).toBeInstanceOf(DOMException);
      expect(back.message).toBe(EXAMPLE.message);
      expect(back.name).toBe(EXAMPLE.name);
    });
  });
  describe('toJSONAsync', () => {
    it('supports DOMException', async () => {
      const result = await toJSONAsync(Promise.resolve(EXAMPLE), {
        plugins: [DOMExceptionPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<typeof EXAMPLE>>(result, {
        plugins: [DOMExceptionPlugin],
      });
      expect(back).toBeInstanceOf(DOMException);
      expect(back.message).toBe(EXAMPLE.message);
      expect(back.name).toBe(EXAMPLE.name);
    });
  });
  describe('crossSerialize', () => {
    it('supports DOMException', () => {
      const result = crossSerialize(EXAMPLE, {
        plugins: [DOMExceptionPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports DOMException', () => {
        const result = crossSerialize(EXAMPLE, {
          plugins: [DOMExceptionPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports DOMException', async () => {
      const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [DOMExceptionPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports DOMException', async () => {
        const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
          plugins: [DOMExceptionPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports DOMException', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(EXAMPLE), {
          plugins: [DOMExceptionPlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports DOMException', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(Promise.resolve(EXAMPLE), {
            plugins: [DOMExceptionPlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports DOMException', () => {
      const result = toCrossJSON(EXAMPLE, {
        plugins: [DOMExceptionPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [DOMExceptionPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(DOMException);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports DOMException', async () => {
      const example = Promise.resolve(EXAMPLE);
      const result = await toCrossJSONAsync(example, {
        plugins: [DOMExceptionPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<typeof example>(result, {
        plugins: [DOMExceptionPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(DOMException);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports DOMException', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(Promise.resolve(EXAMPLE), {
          plugins: [DOMExceptionPlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
