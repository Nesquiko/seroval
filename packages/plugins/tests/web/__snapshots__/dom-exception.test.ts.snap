// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DOMException > crossSerialize > scoped > supports DOMException 1`] = `"($R=>$R[0]=new DOMException("This is an example message.","Example"))($R["example"])"`;

exports[`DOMException > crossSerialize > supports DOMException 1`] = `"$R[0]=new DOMException("This is an example message.","Example")"`;

exports[`DOMException > crossSerializeAsync > scoped > supports DOMException 1`] = `"($R=>$R[0]=Promise.resolve($R[1]=new DOMException("This is an example message.","Example")))($R["example"])"`;

exports[`DOMException > crossSerializeAsync > supports DOMException 1`] = `"$R[0]=Promise.resolve($R[1]=new DOMException("This is an example message.","Example"))"`;

exports[`DOMException > crossSerializeStream > scoped > supports DOMException 1`] = `"($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R["example"])"`;

exports[`DOMException > crossSerializeStream > scoped > supports DOMException 2`] = `"($R=>($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new DOMException("This is an example message.","Example")))($R["example"])"`;

exports[`DOMException > crossSerializeStream > supports DOMException 1`] = `"$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))()"`;

exports[`DOMException > crossSerializeStream > supports DOMException 2`] = `"($R[3]=(p,d)=>{p.s(d),p.status="success",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=new DOMException("This is an example message.","Example"))"`;

exports[`DOMException > serialize > supports DOMException 1`] = `"new DOMException("This is an example message.","Example")"`;

exports[`DOMException > serializeAsync > supports DOMException 1`] = `"Promise.resolve(new DOMException("This is an example message.","Example"))"`;

exports[`DOMException > toCrossJSON > supports DOMException 1`] = `"{"t":25,"i":0,"s":{"name":{"t":1,"s":"Example"},"message":{"t":1,"s":"This is an example message."}},"c":"seroval-plugins/web/DOMException"}"`;

exports[`DOMException > toCrossJSONAsync > supports DOMException 1`] = `"{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"name":{"t":1,"s":"Example"},"message":{"t":1,"s":"This is an example message."}},"c":"seroval-plugins/web/DOMException"}}"`;

exports[`DOMException > toCrossJSONStream > supports DOMException 1`] = `"{"t":22,"i":0,"f":{"t":26,"i":1,"s":1}}"`;

exports[`DOMException > toCrossJSONStream > supports DOMException 2`] = `"{"t":23,"i":0,"a":[{"t":26,"i":3,"s":2},{"t":25,"i":2,"s":{"name":{"t":1,"s":"Example"},"message":{"t":1,"s":"This is an example message."}},"c":"seroval-plugins/web/DOMException"}]}"`;

exports[`DOMException > toJSON > supports DOMException 1`] = `"{"t":{"t":25,"i":0,"s":{"name":{"t":1,"s":"Example"},"message":{"t":1,"s":"This is an example message."}},"c":"seroval-plugins/web/DOMException"},"f":47,"m":[]}"`;

exports[`DOMException > toJSONAsync > supports DOMException 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"name":{"t":1,"s":"Example"},"message":{"t":1,"s":"This is an example message."}},"c":"seroval-plugins/web/DOMException"}},"f":47,"m":[]}"`;
